File: irrgarten-java/src/irrgarten/Player.java
package irrgarten;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class Player {

    private static int MAX_WEAPONS = 2;
    private static int MAX_SHIELDS = 3;
    private static int INITIAL_HEALTH = 10;
    private static int HITS2LOSE = 3;

    private String name;
    private char number;
    private float intelligence;
    private float strength;
    private float health;
    private int row;
    private int col;
    private int consecutiveHits;

    private List<Weapon> weapons;
    private List<Shield> shields;

    public Player(char number, float intelligence, float strength) {
        this.number = number;
        this.intelligence = intelligence;
        this.strength = strength;

        consecutiveHits = 0;
        health = INITIAL_HEALTH;
        name = "Player #" + number;
        setPos(-1, -1);

        weapons = new ArrayList();
        shields = new ArrayList();
    }

    public void resurrect() {
        weapons.clear();
        shields.clear();
        health = INITIAL_HEALTH;
        consecutiveHits = 0;
    }

    public int getRow() {
        return row;
    }

    public int getCol() {
        return col;
    }

    public char getNumber() {
        return number;
    }

    public void setPos(int row, int col) {
        this.row = row;
        this.col = col;
    }

    public boolean dead() {
        return health <= 0;
    }

    public Directions move(Directions direction, Directions[] validMoves) {
        if (!Arrays.asList(validMoves).contains(direction) && validMoves.length > 0) {
            return validMoves[0];
        }
        return direction;
    }

    public float attack() {
        return strength + sumWeapons();
    }

    public boolean defend(float receivedAttack) {
        return manageHit(receivedAttack);
    }

    public void receiveReward() {
        int wReward = Dice.weaponsReward();
        int sReward = Dice.shieldsReward();

        for (int i = 0; i < wReward; i++) {
            Weapon wnew = newWeapon();
            receiveWeapon(wnew);
        }
        for (int i = 0; i < sReward; i++) {
            Shield snew = newShield();
            receiveShield(snew);
        }
        
        health += Dice.healthReward();
    }

    @Override
    public String toString() {
        return "Player{" + "name=" + name + ", number=" + number + ", intelligence=" + intelligence + ", strength=" + strength + ", health=" + health + ", row=" + row + ", col=" + col + ", consecutiveHits=" + consecutiveHits + ", weapons=" + weapons + ", shields=" + shields + '}';
    }

    private void receiveWeapon(Weapon w) {
        weapons.removeIf(we -> we.discard());
        
        if (weapons.size() < MAX_WEAPONS)
            weapons.add(w);
    }

    private void receiveShield(Shield s) {
        shields.removeIf(sh -> sh.discard());
        
        if (shields.size() < MAX_SHIELDS)
            shields.add(s);
    }

    private Weapon newWeapon() {
        return new Weapon(Dice.weaponPower(), Dice.usesLeft());
    }

    private Shield newShield() {
        return new Shield(Dice.shieldPower(), Dice.usesLeft());
    }

    private float sumWeapons() {
        float sum = 0;
        for (Weapon w : weapons) {
            sum += w.attack();
        }
        return sum;
    }

    private float sumShields() {
        float sum = 0;
        for (Shield s : shields) {
            sum += s.protect();
        }
        return sum;
    }

    private float defensiveEnergy() {
        return intelligence + sumShields();
    }

    private boolean manageHit(float receivedAttack) {
        float defense = defensiveEnergy();
        if (defense < receivedAttack) {
            gotWounded();
            incConsecutiveHits();
        } else {
            resetHits();
        }

        boolean lose = consecutiveHits == HITS2LOSE || dead();
        if (lose) {
            resetHits();
        }

        return lose;
    }

    private void resetHits() {
        consecutiveHits = 0;
    }

    private void gotWounded() {
        health--;
    }

    private void incConsecutiveHits() {
        consecutiveHits++;
    }
}

File: irrgarten-ruby/player.rb
# encoding: utf-8
require_relative 'shield'
require_relative 'weapon'

module Irrgarten
  class Player
    @@MAX_WEAPONS = 2
    @@MAX_SHIELDS = 3
    @@INITIAL_HEALTH = 10
    @@HITS2LOSE = 3

    def initialize(number, intelligence, strength)
      @name = "Player #" + number.to_s
      @number = number
      @intelligence = intelligence
      @strength = strength
      @health = @@INITIAL_HEALTH
      @consecutive_hits = 0
      set_pos(-1, -1)

      @weapons = []
      @shields = []
    end

    attr_reader :row, :col, :number # (getters)

    def resurrect
      @health = @@INITIAL_HEALTH
      reset_hits
      @weapons = []
      @shields = []
    end

    def set_pos(row, col)
      @row = row
      @col = col
    end

    def dead
      @health <= 0
    end

    def move(direction, valid_moves)
      if !valid_moves.include?(direction) && valid_moves.size > 0
        return valid_moves[0]
      end
      direction
    end

    def attack
      @strength + sum_weapons
    end

    def defend(received_attack)
      manage_hit(received_attack)
    end

    def receive_reward
      wReward = Dice.weapons_reward
      sReward = Dice.shields_reward
      (0...wReward).each do
        receive_weapon(new_weapon)
      end
      (0...sReward).each do
        receive_shield(new_shield)
      end
      @health += Dice.health_reward
    end

    def to_s
      "#{@name.to_s}: {#{@intelligence.to_s}, #{@health.to_s}, #{@strength.to_s}}"
    end

    private

    def receive_weapon(w)
      @weapons.delete_if do |we|
        we.discard
      end

      if @weapons.size < @@MAX_WEAPONS
        @weapons.push(w)
      end
    end

    def receive_shield(s)
      @shields.delete_if do |sh|
        sh.discard
      end

      if @shields.size < @@MAX_SHIELDS
        @shields.push(s)
      end
    end

    def new_weapon
      Weapon.new(Dice.weapon_power, Dice.uses_left)
    end

    def new_shield
      Shield.new(Dice.shield_power, Dice.uses_left)
    end

    def reset_hits
      @consecutive_hits = 0
    end

    def sum_weapons
      sum_weapons = 0
      @weapons.each do |weapon|
        sum_weapons += weapon.attack
      end
      sum_weapons
    end

    def sum_shields
      sum_shields = 0
      @shields.each do |shield|
        sum_shields += shield.protect
      end
      sum_shields
    end

    def defensive_energy
      @intelligence + self.sum_shields
    end

    def manage_hit(received_attack)
      defense = defensive_energy
      if defense < received_attack
        got_wounded
        inc_consecutive_hits
      else
        reset_hits
      end

      lose = @consecutive_hits == @@HITS2LOSE || dead
      if lose
        reset_hits
      end

      lose
    end

    def got_wounded
      @health -= 1
    end

    def inc_consecutive_hits
      @consecutive_hits += 1
    end

  end
end